### PR TITLE
refactor(app): currentOffsetModal only displays latest offsets

### PR DESCRIPTION
**Pull request by jerader**
closes RLAB-200

# Overview

`CurrentOffsetModal` used to display all offsets even duplicates and labware offsets where the vector values are all 0. This PR creates a util `getLatestCurrentOffsets` that filters out the old offsets and the offsets that are 0. Additionally, this PR wires up the LPC cta in `CurrentOffsetModal` as well as keeping the disabled reason the same as in the main LPC CTA button.

# Changelog

-  create `getLatestCurrentOffsets` and test
- add the new util to `CurrentOffsetsModal`
- create the `CurrentOffsetModal` test 
- create `useLPCDisabledReason` hook that is used in `LaunchLabwarePositionCheck` and `CurrentOffsetsModal`, create test for the hook

# Review requests

go through LPC and add some offsets - go through it a few times, the `CurrentOffsetModal` should display only the latest offsets that were created. if there is a reason for the LPC CTA to be disabled, it should also be reflected in the modal

# Risk assessment

low

### CODE REVIEW

File: app/src/organisms/Devices/ProtocolRun/SetupLabware/utils.ts [hunk 1]
```diff
@@ -83,3 +84,26 @@ export function getLabwareSetupItemGroups(
   )
   return { onDeckItems, offDeckItems }
 }
+
+export function getLatestCurrentOffsets(
+  currentOffsets: LabwareOffset[]
+): LabwareOffset[] {
+  const reverseCurrentOffsets = [...currentOffsets].reverse()
```
Comment (b-cooper) [inline app/src/organisms/Devices/ProtocolRun/SetupLabware/utils.ts]: Why are we reversing the array here? We shouldn't be relying on the inherent order of the offsets that we get back from the robot.

File: app/src/organisms/Devices/ProtocolRun/SetupLabware/utils.ts [hunk 2]
```diff
@@ -83,3 +84,26 @@ export function getLabwareSetupItemGroups(
...
+    ...new Set(
+      reverseCurrentOffsets.map(
+        currentOffset => currentOffset.location.slotName
+      )
+    ),
+  ]
+  const latestCurrentOffsets = reverseCurrentOffsets.filter(
+    (currentOffset, index) =>
+      currentOffset.location.slotName === uniqueSlots[index] &&
+      !(
```
Comment (b-cooper) [inline app/src/organisms/Devices/ProtocolRun/SetupLabware/utils.ts]: we've use `isEqual(currentOffset.vector, IDENTITY_VECTOR)` else for this comparison

File: app/src/assets/localization/en/labware_position_check.json [hunk 3]
```diff
@@ -82,5 +82,6 @@
   "labware_position_check_description": "<block>Labware Position Check is a guided workflow that checks every labware on the deck for an added degree of precision in your protocol.</block><block>Labware Position Check first checks tip racks, and then checks all other labware used in your protocol.</block>",
   "all_modules_and_labware_from_protocol": "All modules and labware used in the protocol",
   "module_in_slot": "{{module}} in {{slot}}",
-  "run_labware_position_check": "run labware position check"
+  "run_labware_position_check": "run labware position check",
+  "get_labware_offset_data": "Get Labware Offset Data"
```
Comment (shlokamin) [inline app/src/assets/localization/en/labware_position_check.json:86]: not introduced by this PR, and probably warrants a bigger discussion with the whole app & ui team, but we probably shouldn't be having any sort of formatting (like capitalization) being done inside of these i18n files. i think we should instead always have these json files have lower case text, and hoist the responsibility of any formatting to be done by the consuming component.

will make a note to self to bring this up when everyone is back

File: app/src/organisms/Devices/ProtocolRun/SetupLabware/utils.ts [hunk 4]
```diff
@@ -83,3 +85,36 @@ export function getLabwareSetupItemGroups(
...
+  currentOffsets: LabwareOffset[]
+): LabwareOffset[] {
+  const nonIdentityOffsets = currentOffsets.filter(
+    currentOffset => !isEqual(currentOffset.vector, IDENTITY_VECTOR)
+  )
+
+  const latestCurrentOffsets = nonIdentityOffsets.reduce<LabwareOffset[]>(
+    (acc, offset) => {
+      const previousMatchIndex = acc.findIndex(
+        ao =>
```
Comment (shlokamin) [inline app/src/organisms/Devices/ProtocolRun/SetupLabware/utils.ts]: nit, can we name `ao` something more descriptive? reduce's are sorta tough to read as is

**Review by shlokamin — COMMENTED** [review]
code lgtm, will test on a robot in a sec

**Comment by shlokamin** [comment]
uh so i'm not sure what is going on, but i can't run LPC at all using protocols generated from prod PD on a robot running with sw version `v6.2.0`.

this line in `depricatedLabware.ts` is the culprit, it resolves to `undefined`: 
```js
const labwareDef = labwareDefinitions[currentLabware.definitionUri]
```

the protocol's labware key does not have a `definitionUri` field (it has a `definitionId` field), but the line above is looking for one. should we be using `getInitialLabwareLocation` instead?

there was some change regarding definitionId vs definitionUri that happened recently right @smb2268?

Screen Shot 2022-12-21 at 5 38 34 PM

**Comment by smb2268** [comment]
> the protocol's labware key does not have a `definitionUri` field (it has a `definitionId` field), but the line above is looking for one. should we be using `getInitialLabwareLocation` instead?

Yeah we removed the translation of `definitionUri` to `definitionId` that used to happen in the `schemaV6Adapter` so now we should always be able to check for the `definitionUri` on the labware key that comes back from protocol analysis. Is this looking directly at the protocol file instead of at the analysis?

When @b-cooper and I were pairing on resolving the conflicts between the schemaV6Adapter removal stuff and his new LPC changes, we decided not to invest the time to make sure the old LPC still worked since it won't ever be accessible in a release. So all of the `deprecated` LPC components may have errors

@shlokamin @jerader

**Comment by jerader** [comment]
> When @b-cooper and I were pairing on resolving the conflicts between the schemaV6Adapter removal stuff and his new LPC changes, we decided not to invest the time to make sure the old LPC still worked since it won't ever be accessible in a release. So all of the `deprecated` LPC components may have errors
> 

I can test LPC with the ff turned on with my robot this morning, and if it works I am leaning towards leaving this as is since we will be removing the deprecated components in the q1 release. @shlokamin @smb2268

**Comment by jerader** [comment]
After discussing with Shlok yesterday, it seems that we want to be able to get through LPC on a robot to test this out before merging. However, it does work properly on simulator & emulator.

When attempting to test on a robot that is up to date with `edge`, I ran into numerous null protection errors with regards to modules since my protocol didn't have any modules. Errors appeared in `getCheckSteps`, `getPrepCommands` and `PrepareSpace`. After fixing those errors and restarting my app and robot, my robot got into a weird state where every single protocol failed protocol analysis because they weren't `valid jsons`. After awhile of trying to fix it, I was unsuccessful. Due to this, i ran out of time and was unable to get through LPC. I will also document this in the ticket but please feel free to merge this in when I'm out.

**Review by b-cooper — APPROVED** [review]
✅

**Pull request merged**